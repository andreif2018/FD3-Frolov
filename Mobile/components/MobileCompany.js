import React from 'react';
import PropTypes from 'prop-types';

import MobileClient from './MobileClient';

import './MobileCompany.css';

class MobileCompany extends React.PureComponent {

    static propTypes = {
        name: PropTypes.string.isRequired,
        headers: PropTypes.arrayOf(
            PropTypes.shape({
                code: PropTypes.number.isRequired,
                header: PropTypes.string.isRequired,
            })
        ),
        clients:PropTypes.arrayOf(
            PropTypes.shape({
                id: PropTypes.number.isRequired,
                lastName: PropTypes.string.isRequired,
                firstName: PropTypes.string.isRequired,
                otchestvo: PropTypes.string.isRequired,
                balance: PropTypes.number.isRequired,
                status: PropTypes.bool.isRequired,
            })
        ),
    };

    state = {
        name: this.props.name,
        clients: this.props.clients,
        showState: 0 // show all
    };

    setName1 = () => {
        this.setState({name:'МТС'});
    };

    setName2 = () => {
        this.setState({name:'Velcom'});
    };

    setBalance = (clientId,newBalance) => {
      let changed=false;
      let newClients=[...this.state.clients]; // копия самого массива клиентов
      newClients.forEach( (c,i) => {
        if ( c.id===clientId && c.balance!==newBalance ) {
          let newClient={...c}; // копия хэша изменившегося клиента
          newClient.balance=newBalance;
          newClients[i]=newClient;
          changed=true;
        }
      } );
      if ( changed )
        this.setState({clients:newClients});
    };

    showAll = () => {
        this.setState({showState: 0} ); // show all
    };

    showActive = () => {
        this.setState({showState: 1} ); // show active
    };

    showBlocked = () => {
        this.setState({showState: 2} ); // show blocked
    };

    addClient = () => {

    };

    render() {

        console.log("MobileCompany render");

        var headerLine = this.props.headers.map((element) => <th key={element.code}>{element.header}</th>);/* формирование заголовков таблицы */

        var clientsCode=this.state.clients.map( client =>
            <MobileClient key={client.id} info={client} show={this.state.showState} />
        );


        return (
            <div className='MobileCompany'>
                <input type="button" value="МТС" onClick={this.setName1} />
                <input type="button" value="Velcom" onClick={this.setName2} />
                <div className='MobileCompanyName'>Компания &laquo;{this.state.name}&raquo;</div>
                <hr/>
                <input type="button" value="Все" onClick={this.showAll} />
                <input type="button" value="Активные" onClick={this.showActive} />
                <input type="button" value="Заблокированные" onClick={this.showBlocked} />
                <hr/>
                <table className="MobileCompanyClients">
                    <thead><tr>{headerLine}</tr></thead>
                    <tbody>{clientsCode}</tbody>
                </table>
                <input type="button" value="Добавить клиента" onClick={this.addClient} />
            </div>
        )
            ;

    }

}

export default MobileCompany;