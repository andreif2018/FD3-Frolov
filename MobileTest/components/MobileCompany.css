.MobileCompany {
    width: auto;
    border: 2px solid lightgray;
    padding: 10px;
    text-align: left;
}

input[type='button'] {
    margin: 20px 0 20px 0 ;
}

td, th {
    width: 140px;
    height: 2em;
    border-top: 1px solid darkgrey;
}