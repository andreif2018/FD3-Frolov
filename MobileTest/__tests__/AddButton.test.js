"use strict";

import React from 'react';
import renderer from 'react-test-renderer';

import MobileCompany from '../components/MobileCompany';

let clientsArr = [
    {id:101, lastName:"Иванов", firstName: "Иван", otchestvo: "Иванович", balance: 200, status: true},
    {id:102, lastName:"Сидоров", firstName: "Сидор", otchestvo: "Сидорович", balance: 250, status: true},
    {id:103, lastName:"Петров", firstName: "Петр", otchestvo: "Петрович", balance: 180, status: true},
    {id:104, lastName:"Григорьев", firstName: "Григорий", otchestvo: "Григорьевич", balance: -220, status: false},
];

let headers = [
    {code:1, header: "Фамилия"},
    {code:2, header: "Имя"},
    {code:3, header: "Отчество"},
    {code:4, header: "Баланс"},
    {code:5, header: "Статус"},
    {code:6, header: "Редактировать"},
    {code:7, header: "Удалить"},
];

const component = renderer.create(
    <MobileCompany headers = {headers} clients={clientsArr}/>
);

let componentTree=component.toJSON();

beforeAll(() => {
    expect(componentTree).toMatchSnapshot('FilterAllButton.test.js.snap');
});

describe('работа кнопки "Добавить клиента" в таблицу', () => {
    // найдём в вёрстке компонента саму кнопку
    const buttonElem = component.root.find( el => (el.type==='input' && el.props.value === 'Добавить клиента') );
    test('нажать на кнопку "Добавить клиента" первый раз после загрузки страницы', () => {
        // "нажмём" на неё
        buttonElem.props.onClick();

        // получаем уже изменённый снэпшот
        componentTree=component.toJSON();
        expect(componentTree).toMatchSnapshot('AddButton.test.js.snap');
    });

    test('нажать на кнопку "Добавить клиента" повторно', () => {
        // "нажмём" на кнопку еще раз
        buttonElem.props.onClick();

        // получаем уже изменённый снэпшот
        componentTree=component.toJSON();
        expect(componentTree).toMatchSnapshot('AddButton.test.js.snap');
    });
});