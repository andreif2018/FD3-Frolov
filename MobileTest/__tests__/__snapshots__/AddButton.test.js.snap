// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddButton.test.js.snap 1`] = `
<div
  className="MobileCompany"
>
  <input
    onClick={[Function]}
    type="button"
    value="Все"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Активные"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Заблокированные"
  />
  <hr />
  <table
    className="MobileCompanyClients"
  >
    <thead>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Иванов
        </td>
        <td>
          Иван
        </td>
        <td>
          Иванович
        </td>
        <td>
          200
        </td>
        <td
          className="Active"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          250
        </td>
        <td
          className="Active"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Петров
        </td>
        <td>
          Петр
        </td>
        <td>
          Петрович
        </td>
        <td>
          180
        </td>
        <td
          className="Active"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Григорьев
        </td>
        <td>
          Григорий
        </td>
        <td>
          Григорьевич
        </td>
        <td>
          -220
        </td>
        <td
          className="Blocked"
        >
          blocked
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          <input
            defaultValue=""
            maxLength="20"
            minLength="2"
            required={true}
            type="text"
          />
        </td>
        <td>
          <input
            defaultValue=""
            maxLength="20"
            minLength="2"
            required={true}
            type="text"
          />
        </td>
        <td>
          <input
            defaultValue=""
            maxLength="20"
            minLength="2"
            required={true}
            type="text"
          />
        </td>
        <td>
          <input
            className="Balance"
            defaultValue={0}
            max="10000"
            min="-1000"
            required={true}
            type="number"
          />
        </td>
        <td
          className="Active"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Сохранить"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <input
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
</div>
`;

exports[`AddButtonSecondTime.test.js.snap 1`] = `
<div
  className="MobileCompany"
>
  <input
    onClick={[Function]}
    type="button"
    value="Все"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Активные"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Заблокированные"
  />
  <hr />
  <table
    className="MobileCompanyClients"
  >
    <thead>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Иванов
        </td>
        <td>
          Иван
        </td>
        <td>
          Иванович
        </td>
        <td>
          200
        </td>
        <td
          className="Active"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          250
        </td>
        <td
          className="Active"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Петров
        </td>
        <td>
          Петр
        </td>
        <td>
          Петрович
        </td>
        <td>
          180
        </td>
        <td
          className="Active"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Григорьев
        </td>
        <td>
          Григорий
        </td>
        <td>
          Григорьевич
        </td>
        <td>
          -220
        </td>
        <td
          className="Blocked"
        >
          blocked
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          <input
            defaultValue=""
            maxLength="20"
            minLength="2"
            required={true}
            type="text"
          />
        </td>
        <td>
          <input
            defaultValue=""
            maxLength="20"
            minLength="2"
            required={true}
            type="text"
          />
        </td>
        <td>
          <input
            defaultValue=""
            maxLength="20"
            minLength="2"
            required={true}
            type="text"
          />
        </td>
        <td>
          <input
            className="Balance"
            defaultValue={0}
            max="10000"
            min="-1000"
            required={true}
            type="number"
          />
        </td>
        <td
          className="Active"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Сохранить"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          <input
            defaultValue=""
            maxLength="20"
            minLength="2"
            required={true}
            type="text"
          />
        </td>
        <td>
          <input
            defaultValue=""
            maxLength="20"
            minLength="2"
            required={true}
            type="text"
          />
        </td>
        <td>
          <input
            defaultValue=""
            maxLength="20"
            minLength="2"
            required={true}
            type="text"
          />
        </td>
        <td>
          <input
            className="Balance"
            defaultValue={0}
            max="10000"
            min="-1000"
            required={true}
            type="number"
          />
        </td>
        <td
          className="Active"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Сохранить"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <input
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
</div>
`;

exports[`FilterAllButton.test.js.snap 1`] = `
<div
  className="MobileCompany"
>
  <input
    onClick={[Function]}
    type="button"
    value="Все"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Активные"
  />
  <input
    onClick={[Function]}
    type="button"
    value="Заблокированные"
  />
  <hr />
  <table
    className="MobileCompanyClients"
  >
    <thead>
      <tr>
        <th>
          Фамилия
        </th>
        <th>
          Имя
        </th>
        <th>
          Отчество
        </th>
        <th>
          Баланс
        </th>
        <th>
          Статус
        </th>
        <th>
          Редактировать
        </th>
        <th>
          Удалить
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Иванов
        </td>
        <td>
          Иван
        </td>
        <td>
          Иванович
        </td>
        <td>
          200
        </td>
        <td
          className="Active"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Сидоров
        </td>
        <td>
          Сидор
        </td>
        <td>
          Сидорович
        </td>
        <td>
          250
        </td>
        <td
          className="Active"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Петров
        </td>
        <td>
          Петр
        </td>
        <td>
          Петрович
        </td>
        <td>
          180
        </td>
        <td
          className="Active"
        >
          active
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
      <tr>
        <td>
          Григорьев
        </td>
        <td>
          Григорий
        </td>
        <td>
          Григорьевич
        </td>
        <td>
          -220
        </td>
        <td
          className="Blocked"
        >
          blocked
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Редактировать"
          />
        </td>
        <td>
          <input
            onClick={[Function]}
            type="button"
            value="Удалить"
          />
        </td>
      </tr>
    </tbody>
  </table>
  <input
    onClick={[Function]}
    type="button"
    value="Добавить клиента"
  />
</div>
`;
